### ISSUE TITLE
Investigate why t0_request CouchDB database grows at a rate of 2GB/h

### ISSUE THREAD

**Issue by amaltaro** [dmwm/WMCore#11304]
**Impact of the bug**
T0 WMAgent

**Describe the bug**
It's been a while that I noticed that the `t0_request` CouchDB is getting lots of updates, increasing the CouchDB database size and running database compaction non-stop. The data size is barely changing, but the data volume (including document updates) increases in a rate of 1GB every 30min, as one can see in this grafana plot:
https://monit-grafana.cern.ch/goto/9Uy9uN4Vk?orgId=11

**How to reproduce it**
Not sure

**Expected behavior**
We need to contact the T0 team and gather information on how the T0 agent is updating these documents.

This database contains a very basic workflow structure, which includes the workflow status and its transitions, e.g.:
https://cmsweb.cern.ch/couchdb/t0_request/Express_Run235284_StreamExpressCosmics

and as such, I would not expect more than 10 revisions (actually N statuses) for each workflow, but contrary to that, this example workflow had 230 revisions (updates!).

This needs to be investigated and the T0 agent changed such that it only updates the request document if it actually has any changes.

**Additional context and error message**
This needs to be done sooner than later

**Comment by jhonatanamado** [comment]
Hi Alan, since the couch update, that component is taking more time to update between cycles. The curren production node using `2.0.4.patch1` shows the [following](https://monit-grafana.cern.ch/d/jvwuG8DGz/cms-tier0-production?orgId=11&from=now-7d&to=now&var-Bin=1h&var-Machine=vocms0314&var-src_rse=All&var-activity=All&var-fts=All&var-WMAgentRequest=All&var-CMSJobType=All&var-AgentUrl=All&viewPanel=103) behaviour in the last 48h. 
<img width="1591" alt="Screen Shot 2022-09-29 at 10 12 11 PM" src="https://user-images.githubusercontent.com/50465903/193182164-b49eaf71-54bc-40e1-b1d0-6cee28550f43.png">
NoData (blue bars) means that the cycle is not been updated since 30 min ago.
Now about `Express_Run235284_StreamExpressCosmics`, that wf was processed long time ago (2015). So no clue why that wf had 230 revisions..  I will try to dig what's going on.

**Comment by amaltaro** [comment]
>  NoData (blue bars) means that the cycle is not been updated since 30 min ago.

@jhonatanamado , if you would like our help to follow up on the slow AgentStatusWatcher cycle (fetching WMBS data), as we briefly discussed over Slack, could you please open a new issue.

In this issue, I would like to see why there are so many updates to the T0 request documents in central CouchDB. Would you or @germanfgv know which components/systems update documents in the t0_request database? I guess it's the Tier0Feeder, but there might be other places as well.

Would you also be able to give me a recent workflow that was completed in T0, for:
* Express
* Repack
* PromptReco
?

**Comment by germanfgv** [comment]
I'm only aware of Tier0Feeder. t0wmadatasvc does not use it and other services are maintained by WMCore, so you probably know them better.

About examples of recent completed workflows
- Express: Express_Run359612_StreamHLTMonitor, Express_Run359597_StreamCalibration
- Repack: Repack_Run359499_StreamRPCMON, Repack_Run359499_StreamPhysics
- PromptReco: PromptReco_Run359495_NoBPTX, PromptReco_Run359495_MinimumBias

**Comment by germanfgv** [comment]
@amaltaro any updates on this? what can we do to advance this investigation?

**Comment by amaltaro** [comment]
@germanfgv German, as we just discussed in the T0 meeting, let me explain what the problem is and then you can start collecting information from the Tier0Feeder.

If you look at the CouchDB document for one of the workflows that you provided above, e.g.:
https://cmsweb.cern.ch/couchdb/t0_request/Express_Run359612_StreamHLTMonitor

you can see the following revision number:
```
"_rev":"2863-94dba43388f5bb9bb50717cab58120ad"
```

which means, after the creation of this document, it has been updated 2862 times.

However, the only information that is supposed to change in this document is `RequestStatus` and `RequestTransition`, and given that the length of the latter is 4, I would expect this workflow NOT to have more than 4 revisions (one revision for each state transition).

Bottom line is, something in the Tier0 agent is updating this document even when there is nothing to be updated. OR, it's updating something to the same something (e.g. status from Merge to Merge).

If you manage to investigate and collect further information for this, please let us know. Otherwise I should be able to get to this in the coming week.

**Comment by germanfgv** [comment]
I suspect the culprit is  `AnalyticsDataCollector`.  These are the last few lines from an agent that's no processing new data since 2 weeks ago.

```
2022-10-26 16:22:45,810:140205750740736:INFO:AnalyticsPoller:Getting Job Couch Data ...
2022-10-26 16:22:45,842:140205750740736:INFO:DataCollectAPI:Found 23 requests
2022-10-26 16:22:45,843:140205750740736:INFO:AnalyticsPoller:Getting FWJRJob Couch Data ...
2022-10-26 16:22:45,874:140205750740736:INFO:AnalyticsPoller:Getting Batch Job Data ...
2022-10-26 16:22:45,904:140205750740736:INFO:AnalyticsPoller:Getting Finished Task Data ...
2022-10-26 16:22:46,053:140205750740736:INFO:AnalyticsPoller:Getting DBS PhEDEx upload status ...
2022-10-26 16:22:46,398:140205750740736:INFO:AnalyticsPoller:Getting Local Queue Data ...
2022-10-26 16:22:46,398:140205750740736:INFO:AnalyticsPoller:Combining data from
                                   Job Couch(23),
                                   FWJR(23),
                                   WorkflowsWithSkippedFile(0),
                                   Batch Job(0),
                                   Finished Tasks(277),
                                   Local Queue(0)
                                   Completed workflows(254)..  ...
2022-10-26 16:22:46,398:140205750740736:INFO:AnalyticsPoller:277 requests Data combined,
 uploading request data...
2022-10-26 16:22:46,489:140205750740736:INFO:Tier0Plugin:Updating the status of 276 workflows
2022-10-26 16:23:07,789:140205750740736:INFO:WMStatsWriter:Committing bulk of 100 docs ...
2022-10-26 16:23:11,758:140205750740736:INFO:WMStatsWriter:Committing bulk of 100 docs ...
2022-10-26 16:23:14,929:140205750740736:INFO:WMStatsWriter:Committing bulk of 77 docs ...
2022-10-26 16:23:17,860:140205750740736:INFO:AnalyticsPoller:Request data upload success
 277 request, 
sleep for next cycle
2022-10-26 16:23:20,352:140205750740736:INFO:BaseWorkerThread:AnalyticsPoller took 34.542 secs to execute
2022-10-26 16:33:24,617:140205750740736:INFO:AnalyticsPoller:Getting Job Couch Data ...
2022-10-26 16:33:24,657:140205750740736:INFO:DataCollectAPI:Found 23 requests
2022-10-26 16:33:24,657:140205750740736:INFO:AnalyticsPoller:Getting FWJRJob Couch Data ...
2022-10-26 16:33:24,690:140205750740736:INFO:AnalyticsPoller:Getting Batch Job Data ...
2022-10-26 16:33:24,715:140205750740736:INFO:AnalyticsPoller:Getting Finished Task Data ...
2022-10-26 16:33:24,807:140205750740736:INFO:AnalyticsPoller:Getting DBS PhEDEx upload status ...
2022-10-26 16:33:25,150:140205750740736:INFO:AnalyticsPoller:Getting Local Queue Data ...
2022-10-26 16:33:25,150:140205750740736:INFO:AnalyticsPoller:Combining data from
                                   Job Couch(23),
                                   FWJR(23),
                                   WorkflowsWithSkippedFile(0),
                                   Batch Job(0),
                                   Finished Tasks(277),
                                   Local Queue(0)
                                   Completed workflows(254)..  ...
2022-10-26 16:33:25,151:140205750740736:INFO:AnalyticsPoller:277 requests Data combined,
 uploading request data...
2022-10-26 16:33:25,257:140205750740736:INFO:Tier0Plugin:Updating the status of 276 workflows
2022-10-26 16:33:40,436:140205750740736:INFO:WMStatsWriter:Committing bulk of 100 docs ...
2022-10-26 16:33:44,245:140205750740736:INFO:WMStatsWriter:Committing bulk of 100 docs ...
2022-10-26 16:33:47,448:140205750740736:INFO:WMStatsWriter:Committing bulk of 77 docs ...
2022-10-26 16:33:50,401:140205750740736:INFO:AnalyticsPoller:Request data upload success
 277 request, 
sleep for next cycle
```
As can be seen, it is still commiting hundresds of docs. I'll check this Tier0Plugin and report back.
